Refactor out some reflective constructor calls

Pr0methean opened the pull request:
Should speed up:

- chunk loading with entities
- tree generation
- spawning entities
- writing NBT with list tags

Also removes unused imports globally.

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into removeReflection

Pr0methean committed:
Refactor out some reflective constructor calls

Should improve performance when generating trees or loading saved chunks
that contain entities.

Pr0methean committed:
Remove reflection from CompoundTag.putList()

Pr0methean committed:
Unused import

Pr0methean committed:
Delete TagType.getConstructor()

Pr0methean committed:
Use MethodHandle to cache reflection where possible

Pr0methean committed:
Style fix

Pr0methean committed:
Revert change to LibraryManager (doesn't quite work)

Pr0methean committed:
Work around https://github.com/mojohaus/animal-sniffer/issues/18

Pr0methean committed:
Switch to ConcurrentHashMap

WeakHashMap won't prevent permgen leaks.

Pr0methean committed:
Unused imports

Pr0methean committed:
Test reducing heap to 512MB

Pr0methean committed:
Make entityCtors static

aramperes commented on src/main/java/net/glowstone/GlowWorld.java:
```diff
@@ -1405,48 +1413,54 @@ public double getHumidity(int x, int z) {
 
         if (TNTPrimed.class.isAssignableFrom(clazz)) {
-            entity = new GlowTntPrimed(location, null);
+            return new GlowTntPrimed(location, null);
```
This prevents the events from being called (under this statement) for Primed TNT.

Pr0methean replied:
Refactored in d78caf8ae29bfd9455a351a4f1dd267d2eb88b91 by overloading the GlowTntPrimed constructor so that it no longer needs separate logic.

aramperes commented on src/main/java/net/glowstone/util/nbt/CompoundTag.java:
```diff
@@ -471,23 +471,19 @@ public void putIntArray(String key, int... value) {
-        } catch (ReflectiveOperationException e) {
-            throw new IllegalArgumentException("Unable to create list of type " + type, e);
+    public <V> void putList(String key, TagType type, List<V> value,
+            Function<? super V, ? extends Tag> tagCreator) {
```
Why not store the method reference in the enum?

Pr0methean replied:
That would create type-safety issues, because an enum can't be generic, so I'm creating `put*List` wrappers instead.

Pr0methean committed:
One-arg ctor for GlowTntPrimed, instead of special handling

Pr0methean committed:
Extract put*List

Pr0methean committed:
Javadoc fix

Pr0methean committed:
Unused imports

Pr0methean committed:
Style fixes

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into removeReflection

Pr0methean committed:
Map entity types to ctors in EntityRegistry

Pr0methean committed:
Unused imports

smartboyathome commented on src/main/java/net/glowstone/entity/EntityRegistry.java:
```diff
@@ -159,107 +166,172 @@
-            .put(ZombieVillager.class, GlowZombieVillager.class)
-            .build();
+            ENTITIES;
+    private static ImmutableBiMap.Builder<Class<? extends Entity>, Class<? extends GlowEntity>>
```
It doesn't appear that we're actually taking advantage of the BiMap here, does it actually need to stay a BiMap? Also, since we already have a CustomEntityContext, I think we should create a BuiltinEntityDescriptor that encapsulates this and entityCtors. Doing this should allow us to move these builders into the static block rather than having them dangle off of the class.

smartboyathome commented on src/main/java/net/glowstone/entity/EntityRegistry.java:
```diff
@@ -159,107 +166,172 @@
+     */
+    public static GlowEntity constructEntity(Class<? extends Entity> type, Location location) {
+        if (ENTITY_CTORS.containsKey(type)) {
+            return ENTITY_CTORS.get(type).apply(location);
```
Actually, I just realized from the below line that EntityStore is essentially a factory for entities, with its createEntity() method. We should just use that rather than duplicating that here, then custom entities will be handled the same as builtins.

Pr0methean replied:
Refactored; see https://github.com/Pr0methean/Glowstone/commit/6baa9be7f050dfb50a037da51f5f7613cb8083e7#diff-8769715641781a189890669ce58619a0R173

Pr0methean committed:
Address https://github.com/GlowstoneMC/Glowstone/pull/860#discussion_r173668583

EntityStorage will now be expected to create new entities for
GlowWorld.spawn(*) by deserializing them from an empty compound tag.

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into removeReflection

Pr0methean committed:
Bug fix

mastercoms committed:
Merge branch 'dev' into removeReflection

Pr0methean committed:
Javadoc fix

Pr0methean committed:
pom.xml bugfix
